domain/application/service/application.go

```diff
@@ -323,8 +327,56 @@ func (s *ApplicationService) AddUnits(ctx context.Context, name string, units ..
 // has been removed from mongo. The mongo calls are
 // DestroyMaybeRemove, DestroyWithForce, RemoveWithForce.
 func (s *ApplicationService) DeleteUnit(ctx context.Context, unitName string) error {
-	err := s.st.DeleteUnit(ctx, unitName)
-	return errors.Annotatef(err, "deleting unit %q", unitName)
+	var cleanups []func(context.Context)
+	err := s.st.RunAtomic(ctx, func(ctx domain.AtomicContext) error {
+		var err error
+		cleanups, err = s.deleteUnit(ctx, unitName)
+		return errors.Trace(err)
+	})
+	if err != nil {
+		return errors.Annotatef(err, "deleting unit %q", unitName)
+	}
+	for _, cleanup := range cleanups {
+		cleanup(ctx)
+	}
+	return nil
+}
+
+func (s *ApplicationService) deleteUnit(ctx domain.AtomicContext, unitName string) ([]func(context.Context), error) {
+	// Get unit owned secrets.
+	uris, err := s.st.GetSecretsForOwners(ctx, application.NilApplicationOwners, application.UnitSecretOwners{unitName})
+	if err != nil {
+		return nil, errors.Annotatef(err, "getting unit owned secrets for %q", unitName)
+	}
+	// Delete unit owned secrets.
+	var cleanups []func(context.Context)
+	for _, uri := range uris {
+		s.logger.Debugf("deleting unit %q secret: %s", unitName, uri.ID)
+		cleanup, err := s.secretService.InternalDeleteSecret(ctx, uri, secretservice.DeleteSecretParams{
```

Comment:
Given we already have secretService here, should we implement  GetSecretsForOwners in secret service?

Reply:
yeah, ok. moved the logic across to the secret service